### PR TITLE
Create a custom weak data structure

### PR DESCRIPTION
**Pull request by vasilmkd**
Part of #2634.

### CODE REVIEW

**Comment by vasilmkd** [comment]
`series/3.x`:
```scala
Benchmark                      (cpuTokens)  (size)   Mode  Cnt    Score   Error  Units
ParallelBenchmark.parTraverse        10000    1000  thrpt   20  223.598 ± 4.735  ops/s
ParallelBenchmark.traverse           10000    1000  thrpt   20   33.490 ± 0.029  ops/s
```

`This PR`:
```scala
Benchmark                      (cpuTokens)  (size)   Mode  Cnt    Score    Error  Units
ParallelBenchmark.parTraverse        10000    1000  thrpt   20  235.694 ± 19.625  ops/s
ParallelBenchmark.traverse           10000    1000  thrpt   20   33.527 ±  0.040  ops/s
```

**Comment by vasilmkd** [comment]
@iRevive Can you please help me run the GC benchmarks from #2634?

**Comment by vasilmkd** [comment]
`series/3.x`:
```scala
Benchmark                                                       (cpuTokens)  (size)   Mode  Cnt        Score        Error   Units
ParallelBenchmark.parTraverse                                         10000    1000  thrpt   20      214.351 ±     14.138   ops/s
ParallelBenchmark.parTraverse:·gc.alloc.rate                          10000    1000  thrpt   20     1093.202 ±     72.011  MB/sec
ParallelBenchmark.parTraverse:·gc.alloc.rate.norm                     10000    1000  thrpt   20  5615317.008 ±   1931.591    B/op
ParallelBenchmark.parTraverse:·gc.churn.G1_Eden_Space                 10000    1000  thrpt   20     1103.003 ±    115.265  MB/sec
ParallelBenchmark.parTraverse:·gc.churn.G1_Eden_Space.norm            10000    1000  thrpt   20  5673279.554 ± 539895.754    B/op
ParallelBenchmark.parTraverse:·gc.churn.G1_Survivor_Space             10000    1000  thrpt   20        0.776 ±      0.590  MB/sec
ParallelBenchmark.parTraverse:·gc.churn.G1_Survivor_Space.norm        10000    1000  thrpt   20     3924.894 ±   2848.428    B/op
ParallelBenchmark.parTraverse:·gc.count                               10000    1000  thrpt   20       99.000               counts
ParallelBenchmark.parTraverse:·gc.time                                10000    1000  thrpt   20    18975.000                   ms
```

`This PR`:
```scala
Benchmark                                                       (cpuTokens)  (size)   Mode  Cnt        Score        Error   Units
ParallelBenchmark.parTraverse                                         10000    1000  thrpt   20      255.577 ±      2.841   ops/s
ParallelBenchmark.parTraverse:·gc.alloc.rate                          10000    1000  thrpt   20     1302.656 ±     14.690  MB/sec
ParallelBenchmark.parTraverse:·gc.alloc.rate.norm                     10000    1000  thrpt   20  5612461.632 ±   1118.883    B/op
ParallelBenchmark.parTraverse:·gc.churn.G1_Eden_Space                 10000    1000  thrpt   20     1316.030 ±     38.440  MB/sec
ParallelBenchmark.parTraverse:·gc.churn.G1_Eden_Space.norm            10000    1000  thrpt   20  5671332.189 ± 184749.189    B/op
ParallelBenchmark.parTraverse:·gc.churn.G1_Survivor_Space             10000    1000  thrpt   20        0.535 ±      0.305  MB/sec
ParallelBenchmark.parTraverse:·gc.churn.G1_Survivor_Space.norm        10000    1000  thrpt   20     2300.459 ±   1299.700    B/op
ParallelBenchmark.parTraverse:·gc.count                               10000    1000  thrpt   20      121.000               counts
ParallelBenchmark.parTraverse:·gc.time                                10000    1000  thrpt   20    22302.000                   ms
```

**Comment by vasilmkd** [comment]
`series/3.x`:
```scala
Benchmark                                              (size)   Mode  Cnt      Score     Error   Units
AsyncBenchmark.async                                      100  thrpt   20  16271.734 ± 298.607   ops/s
AsyncBenchmark.async:·gc.alloc.rate                       100  thrpt   20    658.376 ±  12.067  MB/sec
AsyncBenchmark.async:·gc.alloc.rate.norm                  100  thrpt   20  44552.055 ±   2.320    B/op
AsyncBenchmark.async:·gc.churn.G1_Eden_Space              100  thrpt   20    666.631 ±  16.416  MB/sec
AsyncBenchmark.async:·gc.churn.G1_Eden_Space.norm         100  thrpt   20  45110.623 ± 746.439    B/op
AsyncBenchmark.async:·gc.churn.G1_Survivor_Space          100  thrpt   20      0.002 ±   0.001  MB/sec
AsyncBenchmark.async:·gc.churn.G1_Survivor_Space.norm     100  thrpt   20      0.113 ±   0.047    B/op
AsyncBenchmark.async:·gc.count                            100  thrpt   20    473.000            counts
AsyncBenchmark.async:·gc.time                             100  thrpt   20    499.000                ms
```

`This PR`:
```scala
Benchmark                                              (size)   Mode  Cnt      Score     Error   Units
AsyncBenchmark.async                                      100  thrpt   20  16708.553 ± 395.142   ops/s
AsyncBenchmark.async:·gc.alloc.rate                       100  thrpt   20    676.048 ±  15.953  MB/sec
AsyncBenchmark.async:·gc.alloc.rate.norm                  100  thrpt   20  44551.906 ±   2.815    B/op
AsyncBenchmark.async:·gc.churn.G1_Eden_Space              100  thrpt   20    682.136 ±  18.792  MB/sec
AsyncBenchmark.async:·gc.churn.G1_Eden_Space.norm         100  thrpt   20  44952.287 ± 594.847    B/op
AsyncBenchmark.async:·gc.churn.G1_Survivor_Space          100  thrpt   20      0.003 ±   0.002  MB/sec
AsyncBenchmark.async:·gc.churn.G1_Survivor_Space.norm     100  thrpt   20      0.196 ±   0.124    B/op
AsyncBenchmark.async:·gc.count                            100  thrpt   20    484.000            counts
AsyncBenchmark.async:·gc.time                             100  thrpt   20    576.000                ms
```

**Comment by iRevive** [comment]
Sure, I will do benchmarks tomorrow

**Comment by armanbilge** [comment]
@iRevive Vasil's cooked up something nice, we'll try and get a snapshot for you to try :)

File: core/jvm/src/main/scala/cats/effect/unsafe/WeakDataStructure.scala [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+  @tailrec
+  def insert(key: K, value: V): Unit = {
+    val oldEntry = queue.poll().asInstanceOf[WeakEntry[K, V]]
+    val cap = capacity
+    val sz = size
+
+    if (oldEntry ne null) {
+      // There is an expired entry, and we can reuse its index in the data structure.
+      val index = oldEntry.index
```
Comment (djspiewak) [inline core/jvm/src/main/scala/cats/effect/unsafe/WeakDataStructure.scala:42]: This is actually brilliant

**Review by djspiewak — APPROVED** [review]
This is really great. It needs a better name, but honestly really smart

**Comment by armanbilge** [comment]
@djspiewak the latest is in https://github.com/vasilmkd/cats-effect/tree/weak-data-structure-outlandish I think :)

**Comment by vasilmkd** [comment]
Superseded by #2673.

**Pull request closed without merging**